### PR TITLE
Support multi-precision incomplete gamma function

### PR DESCRIPTION
**Pull request by cbehan**
MPFR 4.0.0 has been out for awhile now so it would be nice to make use of more of its special functions. This does so for UpperGamma and it will be trivial to do so for LowerGamma after. The new code is in functions.cpp because TwoArgFunction doesn't have a general evaluate interface yet.

### CODE REVIEW

**Comment by certik** [comment]
Thanks! That looks good to me. @cbehan please ping me once the tests pass.

File: symengine/functions.cpp [hunk 1]
```diff
@@ -3009,6 +3013,18 @@ RCP<const Basic> uppergamma(const RCP<const Basic> &s,
                            mul(pow(x, s), exp(mul(minus_one, x)))),
                        s);
         }
+#ifdef HAVE_SYMENGINE_MPFR
+    } else if (is_a<RealMPFR>(*s) && is_a<RealMPFR>(*x)) {
+        auto s_ = down_cast<const RealMPFR &>(*s).i.get_mpfr_t();
```
Comment (isuruf) [inline symengine/functions.cpp]:
```suggestion
        const auto &s_ = down_cast<const RealMPFR &>(*s).i.get_mpfr_t();
```

File: symengine/functions.cpp [hunk 2]
```diff
@@ -3009,6 +3013,18 @@ RCP<const Basic> uppergamma(const RCP<const Basic> &s,
...
+        auto x_ = down_cast<const RealMPFR &>(*x).i.get_mpfr_t();
```
Comment (isuruf) [inline symengine/functions.cpp]:
```suggestion
        const auto &x_ = down_cast<const RealMPFR &>(*x).i.get_mpfr_t();
```

**Comment by cbehan** [comment]
Surely the CI issue is not related to this patch anymore.

File: symengine/tests/eval/test_eval_mpfr.cpp [hunk 3]
```diff
@@ -197,6 +198,8 @@ TEST_CASE("precision: eval_mpfr", "[eval_mpfr]")
         std::make_tuple(gamma(add(arg1, arg2)), -100, 100),
         std::make_tuple(gamma(arg1), -100, 100),
         std::make_tuple(gamma(add(add(arg1, arg2), arg1)), -100, 100),
+        std::make_tuple(uppergamma(arg1, arg2), 0.97350097883925,
+                        0.97350097883926),
```
Comment (isuruf) [inline symengine/tests/eval/test_eval_mpfr.cpp]: I'm not sure why tests pass when this should fail with MPFR<4.

Comment (certik) [reply symengine/tests/eval/test_eval_mpfr.cpp]: I think we should investigate that ---- perhaps the test is not run properly.

Comment (isuruf) [reply symengine/tests/eval/test_eval_mpfr.cpp]: Looks like `uppergamma(9/4, 2)` is automatically simplified to `(5/4)*exp(-1/4)` and therefore the `uppergamma` eval is not used.

**Comment by isuruf** [comment]
Thanks.

**Pull request merged**